http_call_test.go

```diff
@@ -0,0 +1,220 @@
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+func TestHTTPCallGetBody(t *testing.T) {
+	t.Parallel()
+	var getBodyCount uint32
+	handler := http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		// The "Connection: close" header is turned into a GOAWAY frame by the http2 server.
+		if atomic.LoadUint32(&getBodyCount) == 0 {
+			responseWriter.Header().Add("Connection", "close")
+		}
+		b, _ := io.ReadAll(request.Body)
+		_ = request.Body.Close()
+		_, _ = responseWriter.Write(b)
+	})
+	server := httptest.NewUnstartedServer(handler)
+	server.EnableHTTP2 = true
+	server.StartTLS()
+	t.Cleanup(server.Close)
+
+	bufferPool := newBufferPool()
+	serverURL, _ := url.Parse(server.URL)
+
+	errGetBodyCalled := fmt.Errorf("getBodyCalled")
+	caller := func(size int) error {
+		call := newHTTPCall(
+			context.Background(),
+			server.Client(),
+			serverURL,
+			Spec{StreamType: StreamTypeUnary},
+			http.Header{},
+		)
+
+		getBodyCalled := false
+		call.onRequestSend = func(*http.Request) {
+			getBody := call.request.GetBody
+			call.request.GetBody = func() (io.ReadCloser, error) {
+				getBodyCalled = true
+				rdcloser, err := getBody()
+				if err != nil {
+					t.Log("getBody failed", err)
+					return nil, err
+				}
+				t.Log("getBodyCalled")
+				atomic.AddUint32(&getBodyCount, 1)
+				return rdcloser, nil
+			}
+		}
+		call.SetValidateResponse(func(*http.Response) *Error {
+			return nil
+		})
+
+		buf := bufferPool.Get()
+		buf.Write(make([]byte, size))
+		if err := call.Send(buf); err != nil {
+			return err
+		}
+		bufferPool.Put(buf)
+		if err := call.CloseWrite(); err != nil {
+			return err
+		}
+		body, err := io.ReadAll(call)
+		if err != nil {
+			return err
+		}
+		if len(body) != size {
+			return fmt.Errorf("expected %d bytes, got %d", size, len(body))
+		}
+		if getBodyCalled {
+			return errGetBodyCalled
+		}
+		return nil
+	}
+	type work struct {
+		size int
+		errs chan error
+	}
+	numWorkers := 2
+	workChan := make(chan work)
+	var wg sync.WaitGroup
+	worker := func() {
+		for work := range workChan {
+			work.errs <- caller(work.size)
+		}
+		wg.Done()
+	}
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go worker()
+	}
+
+	for _, size := range []int{512} {
+		for i, gotGetBody := 0, false; !gotGetBody; i++ {
+			errs := make([]chan error, numWorkers)
+			for i := 0; i < numWorkers; i++ {
+				errs[i] = make(chan error, 1)
+				workChan <- work{size: size, errs: errs[i]}
+			}
+
+			t.Log("waiting", i)
+			for _, errChan := range errs {
+				err := <-errChan
+				if errors.Is(err, errGetBodyCalled) {
+					gotGetBody = true
+				} else if err != nil {
+					t.Fatal(err)
+				}
+			}
+		}
+		x := atomic.LoadUint32(&getBodyCount)
+		if x == 0 {
+			t.Fatal("expected getBody to be called at least once")
+		}
+		atomic.StoreUint32(&getBodyCount, 0)
+	}
+	close(workChan)
+	wg.Wait()
+}
+
+func TestHTTPCallRaces(t *testing.T) {
```

Comment:
This isn't a very nice test but did help recreate the race conditions I was seeing in CI. Any suggestions for improvements?